packages/next/src/server/lib/router-server.ts

```diff
@@ -585,13 +585,20 @@ export async function initialize(opts: {
       })
 
       if (opts.dev && developmentBundler && req.url) {
-        const isHMRRequest = req.url.includes('/_next/webpack-hmr')
+        const { basePath } = config
+
+        const isHMRRequestForHostApp =
+          !basePath && req.url.startsWith('/_next/webpack-hmr')
+
+        const isHMRRequestForGuestApp =
+          basePath && pathHasPrefix(req.url, basePath)
```

Comment:
You are correct @ztanner, the new implementation now accurately identifies HMR requests based on specific criteria, including the presence of 'webpack-hmr' in the URL for both scenarios. This ensures that only relevant requests are treated as HMR requests. 🚀 
Also I added support to [assetPrefix](https://nextjs.org/docs/pages/api-reference/next-config-js/assetPrefix) HMR request for non basePath scenario (for the other scenario is already being handled by pathHasPrefix function) since that was the cause of the failing test (as far I read in docs & code assetPrefix rewrites assets urls for CMS use-cases so now it's handled there too)!
Thank you for bringing this to my attention, and I hope the updated logic aligns with the intended functionality 👌